include/libint2/dfbs_generator.h

```diff
@@ -0,0 +1,305 @@
+        assert(primitive_shells[i].O == primitive_shells[j].O);
+        auto shell = Shell(alpha, contr_, primitive_shells[i].O);
+        if (std::find(product_functions.begin(), product_functions.end(),
+                      shell) == product_functions.end())
+          product_functions.emplace_back(shell);
+      }
+    }
+  }
+  return product_functions;
+}
+
+/// @brief creates a set of candidate product shells from a set of primitive
+/// shells
+/// @param primitive_shells set of primitive shells
+/// @return set of candidate product shells
+inline std::vector<Shell> candidate_functions(
+    const std::vector<Shell> &primitive_shells) {
+  return product_functions(primitive_shells);
+}
+
+/// @brief returns a hash map of shell indices to basis function indices
+inline std::vector<size_t> map_shell_to_basis_function(
+    const std::vector<libint2::Shell> &shells) {
+  std::vector<size_t> result;
+  result.reserve(shells.size());
+
+  size_t n = 0;
+  for (auto &&shell : shells) {
+    result.push_back(n);
+    n += shell.size();
+  }
+
+  return result;
+}
+
+/// @brief computes the Coulomb matrix (\mu|rij^{-1}|\nu)  for a set of shells
+/// @param shells set of shells
+/// @return Coulomb matrix
+inline Eigen::MatrixXd compute_coulomb_matrix(
+    const std::vector<Shell> &shells) {
+  const auto n = nbf(shells);
+  Eigen::MatrixXd result = Eigen::MatrixXd::Zero(n, n);
+  using libint2::Engine;
+  Engine engine(libint2::Operator::coulomb, max_nprim(shells), max_l(shells));
+  engine.set(BraKet::xs_xs);
+  engine.set(ScreeningMethod::Conservative);
+  auto shell2bf = map_shell_to_basis_function(shells);
+  const auto &buf = engine.results();
+  for (size_t s1 = 0; s1 != shells.size(); ++s1) {
+    auto bf1 = shell2bf[s1];
+    auto n1 = shells[s1].size();
+    for (size_t s2 = 0; s2 <= s1; ++s2) {
+      auto bf2 = shell2bf[s2];
+      auto n2 = shells[s2].size();
+      engine.compute(shells[s1], shells[s2]);
+      Eigen::Map<const Eigen::MatrixXd> buf_mat(buf[0], n1, n2);
+      result.block(bf1, bf2, n1, n2) = buf_mat;
+      if (s1 != s2) result.block(bf2, bf1, n2, n1) = buf_mat.transpose();
+    }
+  }
+  return result;
+}
+
+/// @brief Sorts a vector of shells by angular momentum
+inline std::vector<std::vector<Shell>> split_by_L(
+    const std::vector<Shell> &shells) {
+  int lmax = max_l(shells);
+  std::vector<std::vector<Shell>> sorted_shells;
+  sorted_shells.resize(lmax + 1);
+  for (auto &&shell : shells) {
+    auto l = shell.contr[0].l;
+    sorted_shells[l].push_back(shell);
+  }
+  return sorted_shells;
+}
+
+/// @brief computes the reduced set of product functions via pivoted Cholesky
+/// decomposition
+/// @param shells set of shells
+/// @param cholesky_threshold threshold for choosing a product function via
+/// pivoted Cholesky decomposition
+/// @return reduced set of product functions
+inline std::vector<Shell> shell_pivoted_cholesky(
+    const std::vector<Shell> shells, const double cholesky_threshold) {
```

Comment:
Review to ensure your don't pass expensive-to-copy args by value ... Here and elsewhere